File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/OverWindow.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.table.api;

import org.apache.flink.annotation.PublicEvolving;
import org.apache.flink.table.expressions.Expression;

import javax.annotation.Nullable;

import java.util.List;
import java.util.Optional;

/**
 * An over window specification.
 *
 * <p>Similar to SQL, over window aggregates compute an aggregate for each input row over a range of
 * its neighboring rows.
 */
@PublicEvolving
public final class OverWindow {

    private final Expression alias;
    private final List<Expression> partitioning;
    private final Expression order;
    private final @Nullable Expression preceding;
    private final @Nullable Expression following;

    OverWindow(
            Expression alias,
            List<Expression> partitionBy,
            Expression orderBy,
            @Nullable Expression preceding,
            @Nullable Expression following) {
        this.alias = alias;
        this.partitioning = partitionBy;
        this.order = orderBy;
        this.preceding = preceding;
        this.following = following;
    }

    public Expression getAlias() {
        return alias;
    }

    public List<Expression> getPartitioning() {
        return partitioning;
    }

    public Expression getOrder() {
        return order;
    }

    public Optional<Expression> getPreceding() {
        return Optional.ofNullable(preceding);
    }

    public Optional<Expression> getFollowing() {
        return Optional.ofNullable(following);
    }
}

File: flink-python/pyflink/examples/table/windowing/over_window.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
################################################################################
import logging
import sys

from pyflink.datastream import StreamExecutionEnvironment
from pyflink.table import StreamTableEnvironment
from pyflink.table.expressions import col, row_interval, CURRENT_ROW
from pyflink.table.window import Over


def tumble_window_demo():
    env = StreamExecutionEnvironment.get_execution_environment()
    env.set_parallelism(1)
    t_env = StreamTableEnvironment.create(stream_execution_environment=env)

    # define the source with watermark definition
    t_env.execute_sql("""
        CREATE TABLE source (
            ts TIMESTAMP(3),
            name STRING,
            price FLOAT,
            watermark FOR ts as ts - INTERVAL '3' SECOND
        ) with (
            'connector' = 'datagen',
            'number-of-rows' = '10'
        )
    """)

    # define the sink
    t_env.execute_sql("""
        CREATE TABLE sink (
            name STRING,
            total_price FLOAT
        ) with (
            'connector' = 'print'
        )
    """)

    table = t_env.from_path("source")

    # define the over window operation
    table = table.over_window(
        Over.partition_by("name")
            .order_by("ts")
            .preceding(row_interval(2))
            .following(CURRENT_ROW)
            .alias('w')) \
        .select(table.name, table.price.max.over(col('w')))

    # submit for execution
    table.execute_insert('sink') \
         .wait()
    # remove .wait if submitting to a remote cluster, refer to
    # https://nightlies.apache.org/flink/flink-docs-stable/docs/dev/python/faq/#wait-for-jobs-to-finish-when-executing-jobs-in-mini-cluster
    # for more details


if __name__ == '__main__':
    logging.basicConfig(stream=sys.stdout, level=logging.INFO, format="%(message)s")

    tumble_window_demo()
